lib/UR/Observer.pm

```diff
@@ -40,68 +37,106 @@ sub __define__ {
-    if ($@) {
-        $class->error_message("Can't create observer with subject_class_name '$subject_class_name': Can't get class metadata for class '$subject_class_name': $@");
-        return;
-    }
-    unless ($subject_class_meta) {
-        $class->error_message("Class $subject_class_name cannot be the subject class for an observer because there is no class metadata");
-        return;
-    }
 
-    my $aspect = $rule->value_for('aspect');
-    my $subject_id = $rule->value_for('subject_id');
-    unless ($subject_class_meta->_is_valid_signal($aspect)) {
-        if ($subject_class_name->can('validate_subscription') and ! $subject_class_name->validate_subscription($aspect, $subject_id, $callback)) {
-            $class->error_message("'$aspect' is not a valid aspect for class $subject_class_name");
-            return;
+{
+    my @has_defaults = qw(aspect note once priority subject_class_name subject_id);
+    my %defaults =
+        map {
+            $_ => __PACKAGE__->__meta__->{has}->{$_}->{default_value}
+        }
+        grep {
+            exists __PACKAGE__->__meta__->{has}->{$_}
+            && exists __PACKAGE__->__meta__->{has}->{$_}->{default_value}
+        } @has_defaults;
+    sub register_callback {
+        my $class = shift;
+        my %params = @_;
+
+        unless (defined $params{id}) {
+            $params{id} = UR::Object::Type->autogenerate_new_object_id_uuid;
         }
-    }
 
-    if (!defined($subject_class_name) or $subject_class_name eq 'UR::Object') { $subject_class_name = '' }; # This was part of the old API, not sure why it's still here?!
-    if (!defined ($aspect)) { $aspect = '' };
-    if (!defined ($subject_id)) { $subject_id = '' };
-
-    my $self;
-    if ($method eq 'create') {
-        $self = $class->SUPER::create($rule);
-    } elsif ($method eq '__define__') {
-        $self = $class->SUPER::__define__($rule->params_list);
-    } else {
-        Carp::croak('Instantiating a UR::Observer with some method other than create() or __define__() is not supported');
-    }
-    $self->{callback} = $callback;
+        my @missing_params = grep { not exists $params{$_} } @required_params_for_register;
+        if (@missing_params) {
+            Carp::croak('missing required params: ' . join(', ', @missing_params));
+        }
 
-    my %params = $rule->params_list;
-    my ($subscription, $delete_subscription);
+        my @undef_params = grep { not defined $params{$_} } @required_params_for_register;
+        if (@undef_params) {
+            Carp::croak('undefined params: ' . join(', ', @undef_params));
+        }
 
-    $self->_insert_record_into_all_change_subscriptions($subject_class_name, $aspect, $subject_id,
-                                                        [$callback, $self->note, $self->priority, $self->id, $self->once]);
+        my %values = %defaults;
+        %values = map { $_ => delete $params{$_} } @required_params_for_register;
```

Comment:
This second assignment to `%values` replaces the previous one, leaving the `%defaults` hash unused.